### PR TITLE
Feature/ch14894/controller implement the image controller

### PR DESCRIPTION
**Pull request by kentwelcome**
** PR checklist **

- [ ] Ensure you have added or ran the appropriate tests for your PR.
- [ ] DCO signed

**What type of PR is this?**
Feature

**What this PR does / why we need it**:
Implement the custom image builder feature

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
```

### CODE REVIEW

File: controllers/image_controller.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+		}
+
+		// Cancel if cancel flag is marked in Image
+		if image.Spec.ImageSpec.Cancel == true && image.Status.JobCondiction.Phase != CustomImageJobStatusCanceled {
+			return cancel, imageSpecJob
+		}
+
+		// Rebuild if image phase was succeed or failed and image.updateTime after imageSpecJob.updateTime
+		phase := image.Status.JobCondiction.Phase
+		if (phase == CustomImageJobStatusSucceeded || phase == CustomImageJobStatusFailed) &&
```
Comment (popcornylu) [inline controllers/image_controller.go:53]: don't need to check the phase here

**Review by popcornylu — COMMENTED** [review]
review in zoom

**Comment by popcornylu** [comment]
Please add in primehub chart
1. Image additional field
2. Controller RBAC

File: controllers/image_controller.go (hunk 1 above)
Comment (kentwelcome) [reply controllers/image_controller.go]: Removed

File: controllers/image_controller.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+	if image != nil && image.Spec.ImageSpec.BaseImage != "" {
+		// Get ImageSpecJob by Image
+		imageSpecJob := &v1alpha1.ImageSpecJob{}
+		imageSpecJobName := getImageSpecJobName(image)
+		err := r.Get(ctx, client.ObjectKey{Namespace: image.Namespace, Name: imageSpecJobName}, imageSpecJob)
+
+		// Create if no ImageSpecJob found
+		if apierrors.IsNotFound(err) && image.Spec.ImageSpec.Cancel == false {
+			return create, nil
+		}
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply controllers/image_controller.go]: added

File: controllers/image_controller.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+	hash := computeHash(image.Spec.ImageSpec.BaseImage, image.Spec.ImageSpec.Packages)
+
+	imageSpecJob := &v1alpha1.ImageSpecJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        getImageSpecJobName(image),
+			Namespace:   image.Namespace,
+			Annotations: map[string]string{"image.primehub.io/hash": hash},
+			Labels: map[string]string{
+				"image.primehub.io/name":  image.ObjectMeta.Name,
+				"image.primehub.io/group": image.Spec.GroupName,
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply controllers/image_controller.go]: remove groupName and change app to `primehub-image`

File: controllers/imagespec_controller.go [hunk 4]
```diff
@@ -21,6 +21,8 @@ import (
 
 const (
 	CustomImageJobStatusSucceeded = "Succeeded"
+	CustomImageJobStatusCanceled  = "Canceled"
+	CustomImageJobStatusFailed    = "Failed"
 )
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply controllers/imagespec_controller.go]: Move to imagespecjob_controller.go

File: controllers/image_controller.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+	}
+
+	if imageSpecJob != nil {
+		imageClone := image.DeepCopy()
+		imageClone.Status.JobCondiction.JobName = imageSpecJob.Name
+		imageClone.Status.JobCondiction.Phase = imageSpecJob.Status.Phase
+		if imageClone.Status.JobCondiction.Phase == CustomImageJobStatusSucceeded {
+			url := imageSpecJob.Spec.RepoPrefix + "/" + imageSpecJob.Spec.TargetImage
+			imageClone.Status.JobCondiction.Image = url
+			imageClone.Spec.Url = url
```
[the remark this answers is not in the document]
Comment (kentwelcome) [reply controllers/image_controller.go:174]: Added

**Comment by kentwelcome** [comment]
Test image: 
infuseai/primehub-controller-ee:ch14894-e425e6c

**Comment by kentwelcome** [comment]
PrimeHub PR for RBAC and CRD 
https://github.com/InfuseAI/primehub/pull/261

**Comment by kentwelcome** [comment]
Latest test image:
infuseai/primehub-controller-ee:ch14894-8829780

File: controllers/image_controller.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+			Namespace:   image.Namespace,
+			Annotations: map[string]string{"image.primehub.io/hash": hash},
+			Labels: map[string]string{
+				"image.primehub.io/name": image.ObjectMeta.Name,
+				"app":                    "primehub-image",
+			},
+		},
+		Spec: v1alpha1.ImageSpecJobSpec{
+			BaseImage:   image.Spec.ImageSpec.BaseImage,
+			PullSecret:  image.Spec.PullSecret,
```
Comment (popcornylu) [inline controllers/image_controller.go:113]: Copy wrong field. Should by `image.ImageSpec.PullSecret`

File: controllers/image_controller.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+	if imageSpecJob == nil {
+		return fmt.Errorf("cancelImageSpecJob do not provide correct imageSpecJob")
+	}
+
+	if err := r.Delete(ctx, imageSpecJob, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+		return err
+	}
+
+	imageClone := image.DeepCopy()
+	imageClone.Status.JobCondiction.Phase = CustomImageJobStatusCancelled
```
Comment (popcornylu) [inline controllers/image_controller.go:145]: Currently, the cancelled phase would be overwrite. Could we always mark the Phase as `Cancelled` if the `spec.imageSpec.cancel` is true?

**Review by popcornylu — COMMENTED** [review]
Two issues found

File: controllers/image_controller.go (hunk 6 above)
Comment (kentwelcome) [reply controllers/image_controller.go]: Fixed

File: controllers/image_controller.go (hunk 7 above)
Comment (kentwelcome) [reply controllers/image_controller.go:145]: Add logic to check if the image phase is not cancelled but the image had been marked cancelled.

**Comment by kentwelcome** [comment]
Latest test image:
infuseai/primehub-controller-ee:ch14894-a5a6684

**Comment by popcornylu** [comment]
Cancel issue is still not resolved in `ch14894-a5a6684`.
I mark the cancel as true, even the imagespecjob is deleted, the phase is still Running

```
apiVersion: primehub.io/v1alpha1
kind: Image
metadata:
  creationTimestamp: "2021-01-28T15:37:35Z"
  generation: 15
  name: test-xxxx
  namespace: hub
  resourceVersion: "54248709"
  selfLink: /apis/primehub.io/v1alpha1/namespaces/hub/images/test-xxxx
  uid: fa6f9d14-d92a-4910-985e-172f280cf577
spec:
  description: test
  displayName: build custom
  groupName: phusers
  imageSpec:
    baseImage: jupyter/base-notebook:e255f1aa00b2
    cancel: true
    packages:
      apt:
      - vim
    pullSecret: image-gitlab
    updateTime: "2020-12-17T12:44:40Z"
  pullSecret: primehub-controller-custom-image-push-secret
  type: both
  url: gcr.io/infuseai-dev/test-xxxx:84dbd775
  urlForGpu: gcr.io/infuseai-dev/test-xxxx:84dbd775
status:
  jobCondition:
    image: gcr.io/infuseai-dev/test-xxxx:84dbd775
    jobName: test-xxxx
    phase: Running
```

**Comment by kentwelcome** [comment]
Test Image
ch14894-0c6e268

**Comment by popcornylu** [comment]
LGTM

**Pull request merged**